crates/engine/tree/src/tree/mod.rs

```diff
@@ -526,22 +525,25 @@ where
     ///
     /// If the block count exceeds the configured batch size we're allowed to execute at once, this
     /// will execute the first batch and send the remaining blocks back through the channel so that
-    /// don't block request processing for a long time.
-    fn on_downloaded(&mut self, mut blocks: Vec<SealedBlockWithSenders>) -> Option<TreeEvent> {
+    /// block request processing isn't blocked for a long time.
+    fn on_downloaded(
+        &mut self,
+        mut blocks: Vec<SealedBlockWithSenders>,
+    ) -> Result<Option<TreeEvent>, InsertBlockFatalError> {
         if blocks.is_empty() {
             // nothing to execute
-            return None
+            return Ok(None)
         }
 
         trace!(target: "engine", block_count = %blocks.len(), "received downloaded blocks");
         let batch = self.config.max_execute_block_batch_size().min(blocks.len());
         for block in blocks.drain(..batch) {
-            if let Some(event) = self.on_downloaded_block(block) {
+            if let Ok(Some(event)) = self.on_downloaded_block(block) {
```

Comment:
```suggestion
            if let Some(event) = self.on_downloaded_block(block)? {
```

Reply:
The issue I'm having with this is that `on_downloaded_block` can return an InsertBlockFatalError but it doesn't return an option on success and as such the compiler won't let me use `?` operator inside of `run`.